File: demo22/src/demo/scripts/dynamic_pub.py
import rospy
from turtlesim.msg import Pose
import tf2_ros
from geometry_msgs.msg import TransformStamped
import tf_conversions

def doPose(pose):
    # 1.创建一个坐标系相对关系的对象
    pub = tf2_ros.TransformBroadcaster()
    # 2.转换坐标信息
    ts = TransformStamped()
    ts.header.frame_id = "world"
    ts.header.stamp = rospy.Time.now()
    ts.child_frame_id = "turtle1"

    ts.transform.translation.x = pose.x
    ts.transform.translation.y = pose.y
    ts.transform.translation.z = 0

    qtn = tf_conversions.transformations.quaternion_from_euler(0, 0, pose.theta)
    ts.transform.rotation.x = qtn[0]
    ts.transform.rotation.y = qtn[1]
    ts.transform.rotation.z = qtn[2]
    ts.transform.rotation.w = qtn[3]

    # 3.发布
    pub.sendTransform(ts)

if __name__ == "__main__":
    # 1.初始化节点
    rospy.init_node("pub_node")
    # 2.订阅乌龟的位姿信息
    sub = rospy.Subscriber("/turtle1/pose", Pose, doPose, queue_size=100)
    # 3.将位姿信息转化为相对坐标系 callback function

    rospy.spin()

File: demo22/src/demo/src/dynamic_pub.cpp
#include "ros/ros.h"
#include "turtlesim/Pose.h"
#include "tf2_ros/transform_broadcaster.h"
#include "geometry_msgs/TransformStamped.h" // 坐标系转换头文件
#include "tf2/LinearMath/Quaternion.h"      // 四元数转化头文件

void doPose(const turtlesim::Pose::ConstPtr &pose){
    // 1.创建发布对象
    static tf2_ros::TransformBroadcaster pub;       // 这样使用回调函数使用的就是同一个对象
    // 2.组织发布数据
    geometry_msgs::TransformStamped ts;
    ts.header.frame_id = "world";       // 相对坐标系的 id
    ts.header.stamp = ros::Time::now();
    ts.child_frame_id = "turtle1";      // 需要转化的坐标系 id
    // 坐标系偏移量
    ts.transform.translation.x = pose->x;
    ts.transform.translation.y = pose->y;
    ts.transform.translation.z = 0;
    // 坐标系四元数
    /*
        pose 中没有四元数，但有一个theat是偏航角度，同时乌龟是2D的，那么俯仰和翻滚也是0 
    */
    tf2::Quaternion que;
    que.setRPY(0, 0, pose->theta);          // 设置需要转换的坐标系
    ts.transform.rotation.w = que.getW();
    ts.transform.rotation.y = que.getY();
    ts.transform.rotation.z = que.getZ();
    ts.transform.rotation.x = que.getX();
    // 3.发布 
    pub.sendTransform(ts);
}



int main(int argc, char *argv[]){

    // 1.初始化节点
    ros::init(argc, argv, "pub_node");
    // 2.获得句柄
    ros::NodeHandle hd;

    // 3.订阅乌龟的信息
    ros::Subscriber sub = hd.subscribe<const turtlesim::Pose::ConstPtr&>("turtle1/pose", 100, doPose);

    ros::spin();
    return 0;
}
